library/Booster/Pattern/Base.hs

```diff
@@ -396,6 +404,64 @@ externaliseKList def heads optRest
     concatList xs =
         foldr1 (\a b -> SymbolApplication concatSym [] [a, b]) xs
 
+internaliseKSet :: KSetDefinition -> Term -> Term
+internaliseKSet def = \case
+    SymbolApplication s _ []
+        | s.name == def.symbolNames.unitSymbolName -> KSet def [] Nothing
+    SymbolApplication s _ [x]
+        | s.name == def.symbolNames.elementSymbolName -> KSet def [x] Nothing
+    SymbolApplication concatSym _ [x, y]
+        | concatSym.name == def.symbolNames.concatSymbolName ->
+            case (internaliseKSet def x, internaliseKSet def y) of
+                (KSet def1 elems1 rest1, KSet def2 elems2 rest2)
+                    | def1 /= def2 ->
+                        error $ "Inconsistent set definition " <> show (def1, def2)
+                    | def1 /= def ->
+                        error $ "Inconsistent set definition " <> show (def1, def)
+                    | Nothing <- rest1 ->
+                        KSet def (List.sort $ elems1 <> elems2) rest2
+                    | Nothing <- rest2 ->
+                        KSet def (List.sort $ elems1 <> elems2) rest1
+                    | Just r1 <- rest1
+                    , Just r2 <- rest2 ->
+                        KSet
+                            def
+                            (List.sort $ elems1 <> elems2)
+                            (Just $ SymbolApplication concatSym [] [r1, r2])
+                (KSet def1 elems1 rest1, other2)
+                    | def1 /= def -> error $ "Inconsistent set definition " <> show (def1, def)
+                    | Nothing <- rest1 ->
+                        KSet def elems1 (Just other2)
+                    | Just r1 <- rest1 ->
+                        KSet def elems1 (Just $ SymbolApplication concatSym [] [r1, other2])
+                (other1, KSet def2 elems2 rest2)
+                    | def2 /= def -> error $ "Inconsistent set definition " <> show (def2, def)
+                    | Nothing <- rest2 ->
+                        KSet def elems2 (Just other1)
+                    | Just r2 <- rest2 ->
+                        KSet def elems2 (Just $ SymbolApplication concatSym [] [other1, r2])
```

Comment:
Won't this cause looping?? In the other cases for map/list we have something like:
```suggestion
                        KSet def elems2 (Just $ Term attribs $ SymbolApplicationF concatSym [] [other1, r2])
```

Reply:
or perhaps we want to use `stripCollectionMetadata` as in the case below

Reply:
I don' t think it can loop because we are applying concatSym to new, smaller arguments (`r2` resp. `r1` above are strictly subsets of the originally-internalised set). In the case below we might basically reproduce the original term so we need to protect against loops there.
Of course I can protect against the possibility in these two cases as well but I don't think they can loop.